clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -737,16 +751,21 @@ public void handle(SyncGroupResponse syncResponse,
                     log.debug("SyncGroup failed because the group began another rebalance");
                     future.raise(error);
                 } else if (error == Errors.FENCED_INSTANCE_ID) {
-                    log.error("Received fatal exception: group.instance.id gets fenced");
+                    // for sync-group request, even if the generation has changed we would not expect the instance id
+                    // gets fenced, and hence we always treat this as a fatal error
+                    log.error("SyncGroup failed with {} due to group.instance.id {} gets fenced",
+                        sentGeneration, rebalanceConfig.groupInstanceId);
                     future.raise(error);
                 } else if (error == Errors.UNKNOWN_MEMBER_ID
                         || error == Errors.ILLEGAL_GENERATION) {
-                    log.debug("SyncGroup failed: {}", error.message());
-                    resetGenerationOnResponseError(ApiKeys.SYNC_GROUP, error);
+                    log.info("SyncGroup failed with {}: {}, would request re-join", sentGeneration, error.message());
```

Comment:
nit: `would request re-join` or `will request re-join`?

More importantly the log message is outside of the if branch below. But if `generationUnchanged()` is false, then re-join won't be requested, at least not by calling `resetGenerationOnResponseError` -> `resetGeneration` -> `rejoinNeeded = true;`. a) is that safe (as I ask below) b) is the log message printed accurate if there's a chance that re-join is not requested (accuracy of log messages is very crucial here, because usually that's all we have when troubleshooting rebalances).

Reply:
Note on the caller `joinGroupIfNeeded`, when the future fails (here, with either illegal generation and unknown member) then the the join-group would be resent. The only thing we skip is the resetting of the generation since it has already changed.

Reply:
Not sure how this comment was changed, but this still doesn't read well for me. 
Still seems that the message should say `... will request re-join"`